### PR TITLE
pv correlation added

### PR DESCRIPTION
**Pull request by tristanstoeber**

### CODE REVIEW

**Review by lepmik — CHANGES_REQUESTED** [review]
Nice job, we try to keep the function names easily understandable and spell them out when possible, would you mind renaming it to something like "population_vector_correlation"? Moreover, seems like it should be possible to vectorize the min_rate procedure as well as using a boolean slice to extract the population vectors?

**Comment by tristanstoeber** [comment]
what do you think about it now?

File: spatial_maps/tests/test_stats.py [hunk 1]
```diff
@@ -1,6 +1,7 @@
 import numpy as np
 import pytest
 import pdb
```
Comment (lepmik) [inline spatial_maps/tests/test_stats.py]: This is not used - remove

File: spatial_maps/stats.py [hunk 2]
```diff
@@ -257,11 +237,12 @@ def population_vector_correlation(rmaps1, rmaps2, min_rate=None):
         for j in range(bins_y):
             xy1 = rmaps1[id_include, i, j]
             xy2 = rmaps2[id_include, i, j]
-            msk = mask_valid[:, i, j]
+            msk = mask_valid[id_include, i, j]
```
Comment (lepmik) [inline spatial_maps/stats.py]: id_include is no longer used

File: spatial_maps/stats.py [hunk 3]
```diff
@@ -179,3 +182,67 @@ def prob_dist_1d(x, bins):
...
+    bool_nan_c0 = np.isnan(c0)
+    bool_nan_c1 = np.isnan(c1)
+
+    mask_invalid = np.logical_or(bool_nan_c0,
+                                 bool_nan_c1)
+    mask_valid = ~mask_invalid
+    return mask_valid
+
+
+def population_vector_correlation(rmaps1, rmaps2, min_rate=None):
```
Comment (lepmik) [inline spatial_maps/stats.py]: min_rate is no longer used

File: spatial_maps/stats.py [hunk 4]
```diff
@@ -179,3 +182,67 @@ def prob_dist_1d(x, bins):
...
+
+    assert rmaps1.shape == rmaps2.shape
+    # correlation coefficient requires at least 2x2 values
+    assert rmaps1.shape[0] > 1
+
+    corr_coeff = np.zeros((bins_x, bins_y))
+    corr_coeff[:] = np.nan
+    
+    # create mask. exclude value if nan in any of the two stacks
+    mask_valid = _mask_in_both_if_any_is_nan(rmaps1, rmaps2)
```
Comment (lepmik) [inline spatial_maps/stats.py]: I think this should be off by default - to notify the user that there are nan in the stack

File: spatial_maps/stats.py [hunk 5]
```diff
@@ -179,3 +182,67 @@ def prob_dist_1d(x, bins):
...
+    rmaps2 : ndarray
+    Array of the shape [n_units, n_bins_dim1, n_bins_dim2]
+
+    Returns
+    -------
+    pop_vec_corr : float
+    Population vector correlation
+    """
+    # make is float
+    rmaps1 = rmaps1.astype(float)
```
Comment (lepmik) [inline spatial_maps/stats.py]: what is the rationale behind this?

File: spatial_maps/stats.py (hunk 4 above)
Comment (lepmik) [reply spatial_maps/stats.py]: maybe put a mask_nan=False as an argument?

File: spatial_maps/stats.py [hunk 6]
```diff
@@ -179,3 +182,67 @@ def prob_dist_1d(x, bins):
 
     H, _ = np.histogram(x, bins=bins, normed=False)
     return (H / len(x)).T
+
+
+def _mask_in_both_if_any_is_nan(c0, c1):
+    """
+    Mask values in both cubes, if it is nan in cube 0 or/and
```
Comment (lepmik) [inline spatial_maps/stats.py]: this is not only for cubes right? seems that it should work on n dimensional arrays

File: spatial_maps/stats.py [hunk 7]
```diff
@@ -179,3 +182,67 @@ def prob_dist_1d(x, bins):
...
+            xy2 = rmaps2[id_include, i, j]
+            msk = mask_valid[id_include, i, j]
+            # just evaluate correlation if there are any values
+            if np.sum(msk) > 2:
+                corr_coeff[i, j] = np.corrcoef(
+                    xy1[msk],
+                    xy2[msk])[0, 1]
+
+    pop_vec_corr = np.nanmean(corr_coeff)
+    return pop_vec_corr
```
Comment (lepmik) [inline spatial_maps/stats.py]: maybe rather return corr_coeff - nice to see the final map, and easy to do the mean if that's what you want

**Review by lepmik — CHANGES_REQUESTED** [review]

File: spatial_maps/tests/test_stats.py (hunk 1 above)
Comment (tristanstoeber) [reply spatial_maps/tests/test_stats.py]: done

File: spatial_maps/stats.py (hunk 2 above)
Comment (tristanstoeber) [reply spatial_maps/stats.py]: sorry, forgotten

File: spatial_maps/stats.py (hunk 3 above)
Comment (tristanstoeber) [reply spatial_maps/stats.py]: done

File: spatial_maps/stats.py (hunk 4 above)
Comment (tristanstoeber) [reply spatial_maps/stats.py]: good point. made it also slightly more concise, dropped the extra function for the nan detection

File: spatial_maps/stats.py (hunk 6 above)
Comment (tristanstoeber) [reply spatial_maps/stats.py]: dropped the whole function

File: spatial_maps/stats.py (hunk 7 above)
Comment (tristanstoeber) [reply spatial_maps/stats.py]: disagree on this one. the mean over individual correlation coefficients is an integral part of the pv correlation.

Comment (lepmik) [reply spatial_maps/stats.py]: vote for an alternative `return_corr_coeff_map=False`

File: spatial_maps/stats.py [hunk 8]
```diff
@@ -179,3 +179,64 @@ def prob_dist_1d(x, bins):
...
+                mask_valid = ~mask_invalid
+                xy1 = xy1[mask_valid]
+                xy2 = xy2[mask_valid]
+
+            corr_coeff_map[i, j] = np.corrcoef(
+                xy1,
+                xy2)[0, 1]
+
+    if return_corr_coeff_map:
+        return corr_coeff_map
```
Comment (lepmik) [inline spatial_maps/stats.py]: should return both pvcorr and corr_coeff_map

**Review by lepmik — CHANGES_REQUESTED** [review]

**Review by lepmik — APPROVED** [review]
great job :+1:

**Pull request merged**